apps/webapp/src/state/swap.ts

```diff
@@ -90,26 +96,22 @@ export const createSwapSlice = (): SliceCreator<SwapSlice> => (set, get) => {
 };
 
 const assembleSwapRequest = async ({ assetIn, amount, assetOut }: SwapSlice) => {
-  if (assetIn?.value.valueView.case !== 'knownAssetId') throw new Error('unknown denom selected');
-  if (!assetIn.value.valueView.value.metadata?.penumbraAssetId)
-    throw new Error('missing metadata for assetIn');
-  if (assetIn.address.addressView.case !== 'decoded')
-    throw new Error('address in view is not decoded');
-  if (!assetIn.address.addressView.value.index) throw new Error('No index for assetIn address');
-  if (assetOut?.penumbraAssetId === undefined) throw new Error('assetOut has no asset id');
+  if (!assetIn) throw new Error('`assetIn` was undefined');
+
+  const addressIndex = getAddressIndex.orThrow('No index for assetIn address')(assetIn.address);
 
   return new TransactionPlannerRequest({
     swaps: [
       {
-        targetAsset: assetOut.penumbraAssetId,
+        targetAsset: getAssetId.orThrow()(assetOut),
         value: {
           amount: toBaseUnit(
             BigNumber(amount),
-            getDisplayDenomExponent(assetIn.value.valueView.value.metadata),
+            getMetadata.pipe(getDisplayDenomExponent).orThrow()(assetIn.value),
```

Comment:
I think we can move all the `pipe`'s to `types/src/getters/value-view.ts` and have a helper function there. If not, I suspect we'll need to use redundant pipes in different places.

Reply:
Good call — [done](https://github.com/penumbra-zone/web/pull/484/commits/fa259d239a8e592e1d2c24e17dcd084cc5fa1d6f).